### ISSUE TITLE
Provide CKEditor 5 Emoji integration

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#6772]
## Description of the need

CKEditor 5 does not have an easy way to insert emoji. You can roughly imitate some emoji support by [modifying the Special Characters button through custom code](https://ckeditor.com/docs/ckeditor5/latest/features/special-characters.html#adding-a-new-category). But that solution is both difficult and not a very good UI (there's no autocomplete or search).

## Proposed solution

CKEditor is working on an all new Emoji plugin in https://github.com/ckeditor/ckeditor5/issues/17361. Once it is completed, I think this would be a good core candidate feature.

## Alternatives that have been considered

This could be done in contrib as well.

## Additional information

Current progress screenshot from CKEditor:

![image](https://github.com/user-attachments/assets/e476e95c-ca09-4594-96bf-85188b8cdebc)

## Draft of feature description for Press Release (1 paragraph at most)

todo

---

Advocate: @quicksketch

**Comment by quicksketch** [comment]
The new emoji plugin is now available: https://ckeditor.com/docs/ckeditor5/latest/features/emoji.html

**Comment by quicksketch** [comment]
It would be a good idea for us to start planning on switching to the new CKEditor loader introduced in v42, which loads plugins via `import` statements, meaning we won't have to bundle everything in a custom build in the future: https://github.com/ckeditor/ckeditor5/issues/17779

However, it also means all existing contrib plugins will stop working. I'd like to figure out if it's possible to support both loading mechanisms at the same time, but that's a separate issue.

**Comment by indigoxela** [comment]
> ... introduced in v42, which loads plugins via import statements...

TBH, I'm not sure if we can ever switch to that approach (in Backdrop 1.x). There's currently only a handful custom CKE5 plugins in contrib (feef, video_filter, openai, ckeditor_inline_image_style), but we don't want to break them.

Loading (bigger) js files individually might be possible in 1.x, though. But it's some effort.

Long term, for B 1.x, we might need a custom build - it depends on when CKE5 drops the dll build support.

**Comment by indigoxela** [comment]
I went ahead and bundled the Emoji plugin in https://github.com/backdrop/backdrop-issues/issues/6788, we can take care of enabling (hook_ckeditor5_plugins...) here.

**Comment by quicksketch** [comment]
Awesome! I merged the latest PR from https://github.com/backdrop/backdrop-issues/issues/6788 so we are unblocked here.

**Comment by quicksketch** [comment]
I'll advocate for this issue, so I've put it into the `milestone candidate - minor` tag.

**Comment by quicksketch** [comment]
I filed an initial PR at https://github.com/backdrop/backdrop/pull/5034

I quickly ran into an issue where the ability to autocomplete an emoji via typing a colon, followed by an emoji name, such as `:smile:`, did not work because autocomplete is dependent on the [Mentions Plugin](https://ckeditor.com/docs/ckeditor5/latest/features/mentions.html).

In my opinion, having autocomplete on emojis is a necessity for emoji use, so I may be in favor of bundling the Mentions plugin as well... even though I don't have much interest in making core directly handle `@mentions`, which seems like a feature that marketing and single-admin sites would be unlikely to use.

**Comment by indigoxela** [comment]
Personally, I'd not assume the Emoji button to be an 80% use-case. It's cool for community forums and the like, but not for business sites, I guess.

**Comment by indigoxela** [comment]
Re mention plugin: here we go, it's bundled in my latest PR: https://github.com/backdrop/backdrop/pull/5035

**Comment by quicksketch** [comment]
Awesome, thanks! Looks like it's only 20KB extra.

**Comment by quicksketch** [comment]
I updated https://github.com/backdrop/backdrop/pull/5034 to now provide autocomplete on emoji.

Reading the [Emoji Plugin documentation](https://ckeditor.com/docs/ckeditor5/latest/features/emoji.html#emoji-source) closely I also found that CKEditor is retrieving a list of all emoji from a remote CDN. For a CMS like Backdrop, this could be problematic, since a strict configuration of the CORS headers may block external HTTP requests. It would also prevent the editor from loading in offline situations.

To work-around that issue, I copied the JSON list of emoji down and put it in the `build` folder. Perhaps the build script should be updated to include this step? But it is so simple and does not require any actual building/compiling I just updated the `UPGRADING.md` instructions. More details in the PR, this is ready for review.

**Comment by indigoxela** [comment]
Awesome! And a quick turnaround, given we're dealing with CKE5. 😜 

> Perhaps the build script should be updated to include this step?

I think, documentation is sufficient here. Emojis (per standard) don't change that often, CKE5 moves quickly. An independent step seems just fine to me. Apparently the standard gets updated [once a year](https://emojipedia.org/unicode-17.0), and then it takes a while until the changes get rolled out in applications (like browsers, operating systems...).

**Comment by indigoxela** [comment]
Just watching the weekly dev meeting: considering the state of `database_utf8mb4_active` makes a lot of sense, actually.

Let's prevent people from shooting in their own foot - emojis, when unsupported by the DB (unlikely as it is) would lead to PDOExceptions otherwise.

**Comment by indigoxela** [comment]
I set the milestone, so this issue shows up in the milestone listing. Technically it's probably ready. Checking for database_utf8mb4_active still seems like a really good idea to me, given emojis may cause PDOExceptions, but that might be a follow-up as well.

We still need some testing and feedback here. It's a little quiet here. 😁 Seems like the people who usually do testing, have no need for an emoji button. 🙊

**Comment by olafgrabienski** [comment]
> We still need some testing and feedback here.

I've had a look at the sandbox of the PR and was able to insert emojis both ways: using the button, and using the autocomplete starting with a colon. Works for me.

And I agree, we should check for `database_utf8mb4_active`.

**Comment by quicksketch** [comment]
My time is constrained and I think that we could/should push off checking for `database_utf8mb4_active` into a follow-up, and go ahead and merge this for 1.31.0.

**Comment by laryn** [comment]
Thanks  @quicksketch, @indigoxela, and @olafgrabienski! Merged into `1.x` for inclusion in `1.31.0`.

**Comment by quicksketch** [comment]
Thank you @laryn!
